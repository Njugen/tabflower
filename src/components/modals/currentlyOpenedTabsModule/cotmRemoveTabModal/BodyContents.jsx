import React, { Component, Fragment } from "react";
import * as validator from "../../../utils/inputValidators";

export default class BodyContents extends Component {
  render() {
    const { data } = this.props;
    const { isObject, isString, isUndefined } = validator;

    if (isUndefined(data))
      return "The information related to this message is missing. There is no contents to show.";
    if (!isObject(data))
      return "The information related to this message is invalid. The contents cannot be viewed.";
    if (isUndefined(data.params))
      return "The parameter(s) provided by the information container is missing. There is no contents to show.";

    if (!isObject(data.params))
      return "The parameter(s) provided by the information container is invalid. The contents cannot be viewed.";
    const { tabInfo } = data.params;

    if (isObject(tabInfo)) {
      const { title } = tabInfo;

      if (isString(title)) {
        return (
          <Fragment>
            <p>
              You are about to close the tab <strong>{title}</strong>. All
              ongoing activities will be interrupted and possibly lost.
            </p>
            <p>Are you sure you want to proceed?</p>
            <p className="small">
              You may reopen this tab through the browser's history feature
              (presuming you have it activated).
            </p>
          </Fragment>
        );
      } else {
        if (isUndefined(title)) {
          return "The targetted tab's title was not provided. Other info about the targetted tab might also be incorrect, therefore no contents nor settings can be modified at the moment";
        } else {
          return "The targetted tab's title was not provided in the correct manner. Other info about the targetted tab might also be incorrect, therefore no contents nor settings can be modified at the moment";
        }
      }
    } else {
      if (isUndefined(tabInfo)) {
        return "The information about the targetted tab is missing. There are no contents nor settings to modify.";
      } else {
        return "The information about the targetted tab was provided in wrong format. Its contents and settings cannot be shown.";
      }
    }
  }
}
