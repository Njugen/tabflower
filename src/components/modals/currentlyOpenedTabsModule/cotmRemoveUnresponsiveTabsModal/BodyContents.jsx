import React, { Component, Fragment } from "react";
import * as validator from "../../../utils/inputValidators";

export default class BodyContents extends Component {
  render() {
    return <p>Go through all opened tabs and remove them from the list</p>;
  }
}
