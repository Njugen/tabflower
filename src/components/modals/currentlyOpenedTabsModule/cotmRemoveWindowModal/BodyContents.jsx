import React, { Component, Fragment } from "react";

export default class BodyContents extends Component {
  render() {
    return (
      <Fragment>
        <p>
          Closing this window will also close all its tabs. All ongoing
          activities on these web pages will be interrupted and possibly lost.
        </p>
        <p>Are you sure you want to proceed?</p>
        <p className="small">
          You may reopen the window with all its tabs intact from the browser's
          history feature (presuming you have it activated).
        </p>
      </Fragment>
    );
  }
}
