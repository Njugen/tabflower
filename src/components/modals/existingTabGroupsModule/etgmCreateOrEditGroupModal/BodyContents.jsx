import React, { Component, Fragment } from "react";
import * as validator from "../../../utils/inputValidators";
import TBTextInput from "./../../../utils/form/tbTextInput";
import TBTextArea from "./../../../utils/form/tbTextArea";
import TBCheckBox from "./../../../utils/form/tbCheckbox";
import TBWindowListInput from "./../../../utils/form/tbWindowListInput";
import { ValidatorError } from "./../../../utils/exceptionsAndHandler";
import TBScheduleListInput from "../../../utils/form/tbScheduleListInput";

export default class BodyContents extends Component {
  verifyProps = () => {
    const {
      isObject,
      isFunction,
      isUndefined,
      isString,
      isBoolean,
    } = validator;

    const { data, fieldErrors, onChange, tabGroupDetails } = this.props;

    if (!isObject(data))
      throw ValidatorError("etgmCreateOrEditGroupModal-err-1");
    if (!isObject(fieldErrors))
      throw ValidatorError("etgmCreateOrEditGroupModal-err-2");
    if (!isFunction(onChange))
      throw ValidatorError("etgmCreateOrEditGroupModal-err-3");
    if (!isObject(tabGroupDetails))
      throw ValidatorError("etgmCreateOrEditGroupModal-err-4");

    if (isObject(data)) {
      const { params } = data;

      if (!isObject(params))
        throw ValidatorError("etgmCreateOrEditGroupModal-err-5");

      const {
        groupName,
        groupCloseAll,
        groupCloseInactiveTabs,
        groupDescription,
        groupDontAskAgain,
        type,
      } = params;

      // If there is a groupName variable, it has to be a string (groupName is optional)
      if (!isUndefined(groupName) && !isString(groupName))
        throw ValidatorError("etgmCreateOrEditGroupModal-err-6");

      // If there is a groupCloseAll variable, it has to be a bool (groupCloseAll is optional)
      if (!isUndefined(groupCloseAll) && !isBoolean(groupCloseAll))
        throw ValidatorError("etgmCreateOrEditGroupModal-err-7");

      // If there is a groupCloseInactiveTabs variable, it has to be a bool (groupCloseInactiveTabs is optional)
      if (
        !isUndefined(groupCloseInactiveTabs) &&
        !isBoolean(groupCloseInactiveTabs)
      )
        throw ValidatorError("etgmCreateOrEditGroupModal-err-8");

      // If there is a groupDescription variable, it has to be a string (groupDescription is optional)
      if (!isUndefined(groupDescription) && !isString(groupDescription))
        throw ValidatorError("etgmCreateOrEditGroupModal-err-9");

      // If there is a groupDontAskAgain variable, it has to be a bool (groupDontAskAgain is optional)
      if (!isUndefined(groupDontAskAgain) && !isBoolean(groupDontAskAgain))
        throw ValidatorError("etgmCreateOrEditGroupModal-err-10");

      // If there is a type variable, it has to be a string (type is optional)
      if (!isUndefined(type) && !isString(type))
        throw ValidatorError("etgmCreateOrEditGroupModal-err-11");
      if (
        type !== "currently-opened" &&
        type !== "existing-group" &&
        type !== "new-group"
      )
        throw ValidatorError("etgmCreateOrEditGroupModal-err-12");
    }
  };

  componentDidUpdate = () => {
    this.verifyProps();
  };

  render() {
    const { isObject } = validator;
    const { data, fieldErrors, onChange, tabGroupDetails } = this.props;

    if (isObject(data)) {
      const { params } = data;

      if (isObject(params)) {
        const {
          groupName: name,
          groupCloseAll: closeAll,
          groupCloseInactiveTabs: closeInactiveTabs,
          groupDescription: description,
          groupDontAskAgain: dontAskAgain,
          type,
        } = params;

        const {
          groupName: nameErr,
          groupDescription: descErr,
          windowCollection: windowErr,
        } = fieldErrors;

        console.log("MOVERFLOW", tabGroupDetails);
        return (
          <Fragment>
            <TBTextInput
              id="groupName"
              warning={nameErr || null}
              label="Group Name"
              maxWidth={true}
              value={name ? name : ""}
              onChange={(id, value) => onChange(id, value, "tabGroupDetails")}
            ></TBTextInput>
            <TBTextArea
              id="groupDescription"
              warning={descErr || null}
              label="Description (max 170 characters)"
              value={description ? description : ""}
              onChange={(id, value) => onChange(id, value, "tabGroupDetails")}
            ></TBTextArea>
            <TBCheckBox
              id="groupCloseAll"
              label="Close everything else before launching this tab group"
              value={closeAll && closeAll === true ? "true" : "false"}
              onToggle={(id, value) => onChange(id, value, "tabGroupDetails")}
            />
            <TBCheckBox
              id="groupCloseInactiveTabs"
              label="Automatically close all unresponsive tabs opened by this tab group"
              value={
                closeInactiveTabs && closeInactiveTabs === true
                  ? "true"
                  : "false"
              }
              onToggle={(id, value) => onChange(id, value, "tabGroupDetails")}
            />
            <TBCheckBox
              id="groupDontAskAgain"
              label="Do not ask for confirmation when launching this tab group"
              value={dontAskAgain && dontAskAgain === true ? "true" : "false"}
              onToggle={(id, value) => onChange(id, value, "tabGroupDetails")}
            />

            {tabGroupDetails && (
              <>
                <TBWindowListInput
                  windowCollection={tabGroupDetails.windowCollection || []}
                  type={type}
                  label={
                    type === "currently-opened"
                      ? "Currently opened windows and tabs"
                      : type === "existing-group"
                      ? "Edit the windows and tabs in this group"
                      : type === "new-group" &&
                        "Add windows or tabs to this new group"
                  }
                  warning={windowErr || null}
                  onModifyList={onChange}
                />
                <TBScheduleListInput
                  id="groupScheduleList"
                  scheduleCollection={tabGroupDetails.groupScheduleList || []}
                  weekdays={[
                    "Monday",
                    "Tuesday",
                    "Wednesday",
                    "Thursday",
                    "Friday",
                    "Saturday",
                    "Sunday",
                    //  "Every Day",
                  ]}
                  onModifyList={onChange}
                />
              </>
            )}
          </Fragment>
        );
      } else {
        return ValidatorError("ETGMCreateNewGroupModal-135").message;
      }
    } else {
      return ValidatorError("ETGMCreateNewGroupModal-136").message;
    }
  }
}
