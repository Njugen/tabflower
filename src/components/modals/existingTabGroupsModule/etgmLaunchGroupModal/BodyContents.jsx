import React, { Component, Fragment } from "react";
import * as validator from "../../../utils/inputValidators";

import TBCheckBox from "../../../utils/form/tbCheckbox";
import * as ExceptionsHandler from "../../../utils/exceptionsAndHandler";

export default class BodyContents extends Component {
  verifyProps = () => {
    const { data, onChange } = this.props;
    const { isObject, isFunction } = validator;

    if (!isObject(data))
      throw ExceptionsHandler.ValidatorError("ETGMLaunchGroupsModal-113");

    if (!isFunction(onChange))
      throw ExceptionsHandler.ValidatorError("ETGMLaunchGroupsModal-114");
  };

  getWindowCollection = () => {
    const { data } = this.props;
    const { isObject, isArray } = validator;
    const emptyCollection = [];

    if (!isObject(data)) return emptyCollection;

    const { params } = data;

    if (!isObject(params)) return emptyCollection;
    const { windowCollection } = params;

    if (
      !isArray(windowCollection) ||
      (isArray(windowCollection) && windowCollection.length === 0)
    )
      return emptyCollection;

    return windowCollection;
  };

  numberOfWindows = () => {
    const windows = this.getWindowCollection();

    return windows.length;
  };

  numberOfTabs = () => {
    const windows = this.getWindowCollection();
    let numberOfTabs = 0;

    if (windows.length === 0) return 0;

    windows.map((window) => {
      numberOfTabs += window.tabs.length;

      return null;
    });
    console.log(numberOfTabs);

    return numberOfTabs;
  };
  /*
        renderWarningMessage()

        Render a warning message to the user, if the number of windows and/or tabs to be opened exceeds the recommended number.
        The recommended number is up to the app developer to decide. 

        Parameters:
        - numberOfWindows (optional, number)
        - numberOfTabs (optional, number)
    */
  renderWarningMessage = (numberOfWindows, numberOfTabs) => {
    const { isNumber } = validator;

    numberOfWindows = isNumber(numberOfWindows)
      ? numberOfWindows
      : "an undefined number of";
    numberOfTabs = isNumber(numberOfTabs)
      ? numberOfTabs
      : "an undefined number of";

    return (
      <p>
        Warning: You are about to launch a group consisting of{" "}
        <strong>{numberOfWindows} windows</strong> and{" "}
        <strong>{numberOfTabs} tabs</strong>. This might stress your computer
        down in the long run, or cause confusion in your work. Are you sure you
        want to launch this group?
      </p>
    );
  };

  /*
        renderLaunchOptions()

        Render a list of checkbox elements, offering the user the possibility to change 
        the behaviour of the launching tab group.
    */
  renderLaunchOptions = () => {
    const { data, onChange } = this.props;
    const { isObject } = validator;

    if (!isObject(data))
      return "Information about the requested tab group was not provided to this modal. Please, contact the developer.";

    const { params } = data;

    if (!isObject(params))
      return "The options could not be loaded because the relevant parameters were not provided with the group data. Please, contact the developer.";

    const { groupCloseAll, groupCloseInactiveTabs, groupDontAskAgain } = params;

    return (
      <Fragment>
        <TBCheckBox
          id="groupCloseAll"
          label="Close all currently opened tabs and windows"
          value={groupCloseAll && groupCloseAll === true ? "true" : "false"}
          onToggle={(id, value) => onChange(id, value, "tabGroupDetails")}
        />
        <TBCheckBox
          id="groupCloseInactiveTabs"
          label="Automatically close all unresponsive tabs opened by this tab group"
          value={
            groupCloseInactiveTabs && groupCloseInactiveTabs === true
              ? "true"
              : "false"
          }
          onToggle={(id, value) => onChange(id, value, "tabGroupDetails")}
        />
        <TBCheckBox
          id="groupDontAskAgain"
          label="Save all selected options and do not show this message again (All settings offered in this popup can still be changed for any tab group. Just click the cog wheel for the tab group you want to change)."
          value={
            groupDontAskAgain && groupDontAskAgain === true ? "true" : "false"
          }
          onToggle={(id, value) => onChange(id, value, "tabGroupDetails")}
        />
      </Fragment>
    );
  };

  render() {
    const numberOfWindows = this.numberOfWindows();
    const numberOfTabs = this.numberOfTabs();

    return (
      <Fragment>
        {this.renderLaunchOptions()}
        {(numberOfWindows > 5 || numberOfTabs > 15) &&
          this.renderWarningMessage(numberOfWindows, numberOfTabs)}
      </Fragment>
    );
  }
}
