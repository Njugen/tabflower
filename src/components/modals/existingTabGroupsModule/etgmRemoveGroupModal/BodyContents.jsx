import React, { Component, Fragment } from "react";
import * as validator from "../../../utils/inputValidators";

export default class BodyContents extends Component {
  render() {
    const { isObject } = validator;
    const { data } = this.props;

    if (!isObject(data)) return "ETGMRemoveGroupsModal-105";
    const { params } = data;

    if (!isObject(params)) return "ETGMRemoveGroupsModal-104";
    const { groupId, groupName } = params;

    return (
      <Fragment>
        {!groupId && (
          <p>
            Are you sure you want to remove all existing groups? You will lose
            all saved windows and tabs. This cannot be undone.
          </p>
        )}
        {groupId && (
          <p>
            Are you sure you want to remove the tab group{" "}
            {groupName ? <strong>{groupName}</strong> : ""}? You will lose all
            windows and tabs saved in it. This cannot be undone.
          </p>
        )}
      </Fragment>
    );
  }
}
