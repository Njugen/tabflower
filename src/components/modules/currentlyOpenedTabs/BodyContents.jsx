import React, { Component } from "react";
import AppContext from "../../contexts/AppContextProvider";
import { ValidatorError, ErrorHandler } from "../../utils/exceptionsAndHandler";
import * as validator from "../../utils/inputValidators";
import { sendToBackground } from "../../../services/webextension/APIBridge";

import WindowsList from "./../../utils/windowsList";
require("../../../../node_modules/@fortawesome/fontawesome-free/css/all.min.css");

export default class BodyContents extends Component {
  static contextType = AppContext;

  sendToModal = this.context.sendToModal;

  render() {
    const { ...props } = this.props;
    return (
      <>
        <p>
          Tab Flower offers you an oversight over all opened tabs, making it
          easier to remove tabs and windows you seldom use. You may also scan
          for unresponsive websites and remove them.
        </p>
        <WindowsList
          onRaiseToModal={(data) => this.sendToModal(data)}
          onRaiseToErrorOverlay={(data) => this.sendToErrorOverlay(data)}
          canCloseItems={true}
          initialShowTabs={true}
          initialTabStyle="horizontal"
          {...props}
        />
      </>
    );
  }
}
