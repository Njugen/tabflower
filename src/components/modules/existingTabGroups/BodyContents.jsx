import React, { Component } from "react";
import AppContext from "../../contexts/AppContextProvider";
import * as validator from "../../utils/inputValidators";
import TabGroupItem from "../../utils/tabGroupItem";

require("../../../../node_modules/@fortawesome/fontawesome-free/css/all.min.css");

export default class BodyContents extends Component {
  static contextType = AppContext;

  sendToModal = this.context.sendToModal;

  renderTabGroups = (groups) => {
    groups = groups || [];
    const { ...props } = this.props;
    const { isPositiveNumber } = validator;
    console.log("EXISTING", this.props);
    if (isPositiveNumber(groups.length)) {
      return groups.map((group, i) => {
        return (
          <>
            <TabGroupItem info={group} key={"tg-" + i} {...props} />
          </>
        );
      });
    } else {
      return (
        <div className="no-tab-groups-msg">
          There are no saved tab groups at the moment. You may add a new tab
          group by clicking the "Create a new group" button below
        </div>
      );
    }
  };

  render() {
    const { tabgroups } = this.props;

    return (
      <>
        <div className="existing-tab-groups-module">
          <p>
            To keep the browser's tab bar from overflowing, Tab Flower lets you
            group windows and tabs together. You may create new tab groups at
            any time, and launch their tabs in a categorized manner later.
          </p>
          <div className="existing-tab-groups-list">
            {this.renderTabGroups(tabgroups)}
          </div>
        </div>
      </>
    );
  }
}
