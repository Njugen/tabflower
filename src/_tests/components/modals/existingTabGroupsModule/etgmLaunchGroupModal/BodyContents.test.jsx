import React, { Fragment } from "react";
import { shallow, mount, render } from "enzyme";
import BodyContents from "../../../../../components/modals/existingTabGroupsModule/etgmLaunchGroupModal/BodyContents.jsx";
import * as ExceptionsHandler from "../../../../../components/utils/exceptionsAndHandler";
import * as validator from "../../../../../components/utils/inputValidators";

const predefinedComponent = (props, options) => {
  props = props || {};

  const component = shallow(<BodyContents {...props} />, options);
  component.instance().render = jest.fn();
  return component;
};

let presetProps = {
  data: {},
  onRaiseToErrorOverlay: "",
  onDismiss: "",
};

let testComponent;
let componentInstance;

//jest.mock("./__mocks__/fetch");

const actualErrorReturns = {
  "ETGMLaunchGroupsModal-113": ExceptionsHandler.ValidatorError(
    "ETGMLaunchGroupsModal-113"
  ),
  "ETGMLaunchGroupsModal-114": ExceptionsHandler.ValidatorError(
    "ETGMLaunchGroupsModal-114"
  ),
};

const expectedErrorReturns = {
  "ETGMLaunchGroupsModal-113": {
    name: "ValidatorError",
    message: "The data key in props needs to be an object",
    code: "ETGMLaunchGroupsModal-113",
  },
  "ETGMLaunchGroupsModal-114": {
    name: "ValidatorError",
    message: "The onChange key in props needs to be a function",
    code: "ETGMLaunchGroupsModal-114",
  },
};

beforeEach(() => {
  jest.clearAllMocks();
  jest.useRealTimers();

  const presetProps = {};
  testComponent = predefinedComponent(presetProps, {
    disableLifecycleMethods: true,
  });
  componentInstance = testComponent.instance();

  ExceptionsHandler.ErrorHandler = jest.fn();
  ExceptionsHandler.ValidatorError = jest.fn();
  ExceptionsHandler.ValidatorError.mockImplementation((errCode) => {
    return actualErrorReturns[errCode];
  });
});

const various_nonObjects = [
  ["a very weird looking text string"],
  [77],
  [false],
  [true],
  [undefined],
  [[1, 2, 3, 4]],
  [() => {}],
  [null],
];

const various_nonString = [
  [{ testkey: "test value" }],
  [32],
  [null],
  [undefined],
  [false],
  [true],
  [[12, 8, 3, 7]],
  [() => {}],
];

const various_nonBool = [
  ["a very weird looking text string"],
  [{ testkey: "test value" }],
  [32],
  [null],
  [undefined],
  [[12, 8, 3, 7]],
  [() => {}],
];

const various_nonString_nonUndefined = [
  [{ testkey: "test value" }],
  [32],
  [null],
  [false],
  [true],
  [[12, 8, 3, 7]],
  [() => {}],
];

const various_nonBool_nonUndefined = [
  [{ testkey: "test value" }],
  [32],
  [null],
  [[12, 8, 3, 7]],
  [() => {}],
];

const various_nonFunctions = [
  [{ testkey: "test value" }],
  [32],
  [null],
  [undefined],
  [false],
  [true],
  [[12, 8, 3, 7]],
  ["a text string"],
];

const various_nonArrays = [
  [{ testkey: "test value" }],
  [32],
  [null],
  [undefined],
  [false],
  [true],
  [() => {}],
  ["a text string"],
];

const various_nonArrays_nonUndefined = [
  [{ testkey: "test value" }],
  [32],
  [null],
  [false],
  [true],
  [() => {}],
  ["a text string"],
];

const various_nonNumber = [
  [{ testkey: "test value" }],
  [[12, 8, 3, 7]],
  [null],
  [undefined],
  [false],
  [true],
  [() => {}],
  ["a text string"],
];

describe("Test <BodyContents /> component behaviour at mount", () => {
  describe("Test verifyProps()", () => {
    describe("Examine function behaviour based on the value of this.props.data", () => {
      test('Run verifyProps(): If this.props.data is missing, throw an error "ETGMLaunchGroupsModal-113"', () => {
        const presetProps = {};
        testComponent = predefinedComponent(presetProps, {
          disableLifecycleMethods: true,
        });
        componentInstance = testComponent.instance();

        expect(() => componentInstance.verifyProps()).toThrowError(
          ExceptionsHandler.ValidatorError("ETGMLaunchGroupsModal-113").message
        );
      });

      test.each(various_nonObjects)(
        'Run verifyProps(): If this.props.data = %p (is not an object), throw an error "ETGMLaunchGroupsModal-113"',
        (val) => {
          const presetProps = {
            data: val,
          };
          testComponent = predefinedComponent(presetProps, {
            disableLifecycleMethods: true,
          });
          componentInstance = testComponent.instance();

          expect(() => componentInstance.verifyProps()).toThrowError(
            ExceptionsHandler.ValidatorError("ETGMLaunchGroupsModal-113")
              .message
          );
        }
      );

      test('Run verifyProps(): If this.props.data = {}, do not throw error "ETGMLaunchGroupsModal-113"', () => {
        const presetProps = {
          data: {},
        };
        testComponent = predefinedComponent(presetProps, {
          disableLifecycleMethods: true,
        });
        componentInstance = testComponent.instance();

        expect(() => componentInstance.verifyProps()).not.toThrowError(
          ExceptionsHandler.ValidatorError("ETGMLaunchGroupsModal-113").message
        );
      });
    });

    describe("Examine function behaviour based on the value of this.props.onChange", () => {
      test('Run verifyProps(): If this.props.onChange is missing, throw an error "ETGMLaunchGroupsModal-114"', () => {
        const presetProps = {
          data: {},
        };
        testComponent = predefinedComponent(presetProps, {
          disableLifecycleMethods: true,
        });
        componentInstance = testComponent.instance();

        expect(() => componentInstance.verifyProps()).toThrowError(
          ExceptionsHandler.ValidatorError("ETGMLaunchGroupsModal-114").message
        );
      });

      test.each(various_nonFunctions)(
        'Run verifyProps(): If this.props.onChange = %p (is not a function), throw an error "ETGMLaunchGroupsModal-114"',
        (val) => {
          const presetProps = {
            data: {},
            onChange: val,
          };
          testComponent = predefinedComponent(presetProps, {
            disableLifecycleMethods: true,
          });
          componentInstance = testComponent.instance();

          expect(() => componentInstance.verifyProps()).toThrowError(
            ExceptionsHandler.ValidatorError("ETGMLaunchGroupsModal-114")
              .message
          );
        }
      );

      test('Run verifyProps(): If this.props.onChange is a function, do not throw error "ETGMLaunchGroupsModal-114"', () => {
        const presetProps = {
          data: {},
          onChange: () => {},
        };
        testComponent = predefinedComponent(presetProps, {
          disableLifecycleMethods: true,
        });
        componentInstance = testComponent.instance();

        expect(() => componentInstance.verifyProps()).not.toThrowError(
          ExceptionsHandler.ValidatorError("ETGMLaunchGroupsModal-114").message
        );
      });
    });
  });

  describe("Test getWindowCollection()", () => {
    describe("Examine function behaviour based on the value of this.props.data", () => {
      test("Run getWindowCollection(): If this.props.data is missing, return an empty array", () => {
        const presetProps = {};
        testComponent = predefinedComponent(presetProps, {
          disableLifecycleMethods: true,
        });
        componentInstance = testComponent.instance();

        expect(componentInstance.getWindowCollection()).toStrictEqual([]);
      });

      test.each(various_nonObjects)(
        "Run getWindowCollection(): If this.props.data = %p (not an object), return an empty array",
        (val) => {
          const presetProps = {
            data: val,
          };
          testComponent = predefinedComponent(presetProps, {
            disableLifecycleMethods: true,
          });
          componentInstance = testComponent.instance();

          expect(componentInstance.getWindowCollection()).toStrictEqual([]);
        }
      );
    });
  });

  describe("Test numberOfWindows()", () => {});

  describe("Test numberOfTabs()", () => {});

  describe("Test renderWarningMessage()", () => {});

  describe("Test renderLaunchOptions()", () => {
    describe("Examine the function based on the value of this.props.data", () => {
      test('Run renderLaunchOptions(): If "data" is missing in this.props, return a string informing the user about this', () => {
        const presetProps = {};
        testComponent = predefinedComponent(presetProps, {
          disableLifecycleMethods: true,
        });
        componentInstance = testComponent.instance();

        expect(componentInstance.renderLaunchOptions()).toBe(
          "Information about the requested tab group was not provided to this modal. Please, contact the developer."
        );
      });

      test.each(various_nonObjects)(
        'Run renderLaunchOptions(): If "data" = %p (is not an object) in this.props, return a string informing the user about this',
        (val) => {
          const presetProps = {
            data: val,
          };
          testComponent = predefinedComponent(presetProps, {
            disableLifecycleMethods: true,
          });
          componentInstance = testComponent.instance();

          expect(componentInstance.renderLaunchOptions()).toBe(
            "Information about the requested tab group was not provided to this modal. Please, contact the developer."
          );
        }
      );

      test('Run renderLaunchOptions(): If "data" is an object, do not trigger error handling', () => {
        const presetProps = {
          data: {},
        };
        testComponent = predefinedComponent(presetProps, {
          disableLifecycleMethods: true,
        });
        componentInstance = testComponent.instance();

        expect(componentInstance.renderLaunchOptions()).not.toBe(
          "Information about the requested tab group was not provided to this modal. Please, contact the developer."
        );
      });
    });

    describe("Examine the function based on the value of this.props.data.params", () => {
      test('Run renderLaunchOptions(): If "params" is missing in this.props.data, return a string informing the user about this', () => {
        const presetProps = {
          data: {},
        };
        testComponent = predefinedComponent(presetProps, {
          disableLifecycleMethods: true,
        });
        componentInstance = testComponent.instance();

        expect(componentInstance.renderLaunchOptions()).toBe(
          "The options could not be loaded because the relevant parameters were not provided with the group data. Please, contact the developer."
        );
      });

      test.each(various_nonObjects)(
        'Run renderLaunchOptions(): If "params" = %p (is not an object) in this.props.data, return a string informing the user about this',
        (val) => {
          const presetProps = {
            data: {
              params: val,
            },
          };
          testComponent = predefinedComponent(presetProps, {
            disableLifecycleMethods: true,
          });
          componentInstance = testComponent.instance();

          expect(componentInstance.renderLaunchOptions()).toBe(
            "The options could not be loaded because the relevant parameters were not provided with the group data. Please, contact the developer."
          );
        }
      );

      test('Run renderLaunchOptions(): If the params key in "data" is an object, do not trigger error handling', () => {
        const presetProps = {
          data: {
            params: {},
          },
        };
        testComponent = predefinedComponent(presetProps, {
          disableLifecycleMethods: true,
        });
        componentInstance = testComponent.instance();

        expect(componentInstance.renderLaunchOptions()).not.toBe(
          "The options could not be loaded because the relevant parameters were not provided with the group data. Please, contact the developer."
        );
      });
    });

    describe("Examine the function, when this.props.data and this.props.data.params are valid", () => {
      /* Snapshot testing coming later */
    });
  });
});
