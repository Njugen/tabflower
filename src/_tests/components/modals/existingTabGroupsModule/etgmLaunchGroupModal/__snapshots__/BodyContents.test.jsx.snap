// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <BodyContents /> component behaviour at mount Test renderLaunchOptions() Examine the function, when this.props.data and this.props.data.params are valid Run renderLaunchOptions(): The output based on the preset props should match the snapshot. Evaluate the snapshot after every run of this test case. 1`] = `
<React.Fragment>
  <TBCheckBox
    id="groupCloseAll"
    label="Close all currently opened tabs and windows"
    onToggle={[Function]}
    value="false"
  />
  <TBCheckBox
    id="groupCloseInactiveTabs"
    label="Automatically close all unresponsive tabs opened by this tab group"
    onToggle={[Function]}
    value="true"
  />
  <TBCheckBox
    id="groupDontAskAgain"
    label="Save all selected options and do not show this message again (All settings offered in this popup can still be changed for any tab group. Just click the cog wheel for the tab group you want to change)."
    onToggle={[Function]}
    value="true"
  />
</React.Fragment>
`;
